templates/custom.yml.j2

```diff
@@ -0,0 +1,3 @@
+{% if devture_traefik_configuration_custom_file_enabled %}
+  {{ devture_traefik_configuration_custom_file_content | to_nice_yaml }} 
+{% endif %}
```

Comment:
I suppose installing this configuration as a separate file is okay.

Keep in mind that other roles (especially those in matrix-docker-ansible-deploy) take another approach for extending YAML configuration files - mostly because most pieces of software can only load configuration from a single file.

These other roles contain a `_configuration_extension_yaml` variable (similar to `devture_traefik_configuration_custom_file_content`). This variable's content is parsed as YAML and **merged** with the original configuration in an effort to produce a combined configuration which contains everything.

Here's an example from the `matrix-synapse` role:

- the default Synapse configuration lives in [the `homeserver.yaml.j2` template](https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/master/roles/custom/matrix-synapse/templates/synapse/homeserver.yaml.j2)

- this default Synapse configuration is [loaded into a `matrix_synapse_configuration_yaml` variable here](https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/99f69822877a65d55b42a0755af34e124cea8e53/roles/custom/matrix-synapse/defaults/main.yml#L975-L980)

- users can override the configuration via [the `matrix_synapse_configuration_extension_yaml` variable defined here](https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/99f69822877a65d55b42a0755af34e124cea8e53/roles/custom/matrix-synapse/defaults/main.yml#L982C1-L997)

- the default configuration and the configuration extension are combined together to [produce the final `matrix_synapse_configuration` variable](https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/99f69822877a65d55b42a0755af34e124cea8e53/roles/custom/matrix-synapse/defaults/main.yml#L999-L1003)

- the combined configuration is [put onto a file here](https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/99f69822877a65d55b42a0755af34e124cea8e53/roles/custom/matrix-synapse/tasks/synapse/setup_install.yml#L109-L115)

---------

I suppose that this approach may be better, because:

- `traefik.yml` being in the same `/config` directory may cause trouble
- it's not clear what the order of loading and merging for multiple files (`provider.yml`, `custom.yml`) is for Traefik. If it's alphabetical, then I suppose `custom.yml` would be loaded before `provider.yml` and cannot override things defined in it
- it may be harder for people to reason about configuration when it lives in multiple files. It's probably better if the custom configuration is combined with the default one and produces one good final file that people can inspect (in `/config/provider.yml`) and reason about

Due to these reasons, I believe it would be better to redo this to use a single file and merged the default and extended configurations into it.

Reply:
Hi, 

sorry for the late response.
When merging the conf with a extend-conf we need to make sure to split the blocks correctly, middlewares must stay in middlewares, routers must stay in routers etc...

I had trouble using the conf-extension, looked like traefik need to have one conf-file correctly ordered, you cannot just attach another middlewares-block.

That is why I was building an own conf-file. Please feel free to test that and correct me.

In my configuration the traefik.yml does not make any issues, in the logs it looks like the file is not processed at all.

regards

Reply:
The Ansible `|combine` filter invoked with `recursive=True` is supposed to properly merge all these sections. See here: https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/99f69822877a65d55b42a0755af34e124cea8e53/roles/custom/matrix-synapse/defaults/main.yml#L1003